File: Console/src/io/pipe.ts
/**
 * Copyright (c) 2017-2018 Structured Data, LLC
 * 
 * This file is part of BERT.
 *
 * BERT is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * BERT is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with BERT.  If not, see <http://www.gnu.org/licenses/>.
 */

import * as net from "net";
import * as messages from "../../generated/variable_pb.js";
import * as Rx from "rxjs";

import { MessageUtilities } from '../common/message_utilities';

enum Channel {
  INTERNAL,
  CALL,
  SYSCALL, 
  SHELL,
  SYSTEM
}

interface QueuedCommand {
  command: string;
  channel: Channel;
  resolve: Function;
  reject: Function;
  id?: number;
}

export enum ConsoleMessageType {
  TEXT,
  ERR,
  PROMPT,
  GRAPHICS,
  MIME_DATA
}

export interface ConsoleMessage {
  id:number;
  text?:string;
  data?:any;
  type:ConsoleMessageType;
  mime_type?:string;
}

export interface HistoryCallbackType { (options?:any): Promise<string[]> }

export class Pipe {

  /** id for transactions */
  private transaction_id_ = 1;

  /** current pending transaction(s) */
  private pending_: {[index:number]: QueuedCommand} = {};

  /** queued commands */
  private queue_: QueuedCommand[] = [];

  /** pipe name */
  private pipe_name_ = "";

  /** pipe connection (nodejs socket) */
  private client_: any;

  /** 
   * observable state. we can't know about actual state, this
   * just reflects whether _we_ (the console client) are running 
   * a command.
   */
  private busy_status_ = new Rx.BehaviorSubject<boolean>(false);

  /** accessor */
  public get busy_status(){ return this.busy_status_; }

  /**
   * observable for console messages
   */
  private console_messages_ = new Rx.Subject<ConsoleMessage>();

  /** accessor */
  get console_messages(): Rx.Subject<ConsoleMessage> {
    return this.console_messages_;
  }

  /** graphics messages are special - they may have responses */
  private graphics_message_handler_:Function;

  set graphics_message_handler(handler:Function){
    this.graphics_message_handler_ = handler;
  }

  /** queries history. has a default implementation. */
  private history_callback_:HistoryCallbackType = async (options:any = {}) => [];

  set history_callback(callback:HistoryCallbackType){
    this.history_callback_ = callback;
  }

  /**
   * observable for control messages
   */
  private control_messages_ = new Rx.Subject<string>();

  /** accessor */
  get control_messages(): Rx.Subject<string>{
    return this.control_messages_;
  }

  /** accessor dereferences subject */
  get busy() { 
    return this.busy_status_.value; 
  }

  constructor() {
 
  }

  Internal(command) {
    return this.Queue(command, Channel.INTERNAL);
  }

  ShellExec(command) {
    return this.Queue(command, Channel.SHELL);
  }

  Control(command) {
    return this.Queue(command, Channel.SYSTEM);
  }

  SysCall(command){
    return this.Queue(command, Channel.SYSCALL);
  }

  private Queue(command, channel) : Promise<any> {
    return new Promise((resolve, reject) => {
      this.queue_.push({
        command, channel, resolve, reject
      });
      if (!this.busy) this.ProcessQueue();
    });
  }

  private SetBusy(busy) {
    this.busy_status_.next(busy);
  }

  private ProcessQueue() {

    if (this.busy) throw ("busy!");

    if (this.queue_.length === 0) {
      // this.pending_ = null;
      return;
    }

    let message = this.queue_.shift();
    message.id = ++this.transaction_id_;

    // console.info("TX message", message.id, message);

    let call = new messages.CallResponse();
    call.setId(message.id);
    call.setWait(true);

    let function_call;

    switch (message.channel) {
      case Channel.INTERNAL:
        let code = new messages.Code();
        code.setLineList([message.command]);
        call.setCode(code);
        break;
      case Channel.SYSTEM:
        function_call = new messages.CompositeFunctionCall;
        function_call.setTarget(messages.CallTarget.SYSTEM);
        function_call.setFunction(message.command);
        call.setFunctionCall(function_call);
        break;
      case Channel.SYSCALL:
        function_call = new messages.CompositeFunctionCall;
        function_call.setTarget(messages.CallTarget.SYSTEM);
        function_call.setFunction(message.command);
        call.setFunctionCall(function_call);
        break;
      case Channel.CALL:
        break;
      case Channel.SHELL:
        call.setShellCommand(message.command);
        break;
    }

    /*
    // if we are not expecting a response, resolve when the write completes

    let callback = call.getWait() ? null :() => {
      this.SetBusy(false);
      if( this.pending_.resolve) setImmediate(() => this.pending_.resolve(null));
    };
    */

    this.pending_[message.id] = message;
    this.SetBusy(true);

    let data = call.serializeBinary();
    let frame_length = new Int32Array(1);
    let frame = new Uint8Array(data.length + 4);

    frame_length[0] = data.length;
    frame.set(new Uint8Array(frame_length.buffer), 0);
    frame.set(data, 4);

    try {
      this.client_.write(Buffer.from(frame as any)); // ts type is wrong?
    }
    catch(e){
      console.error(e);
      this.FlushQueueOnError("write error");
    }

  }

  /**
   * flush queue and any pending responses. all promises will be rejected.
   * generally this indicates that a pipe/language went down. 
   * 
   * FIXME: support re-establishing languages/pipes/connections. 
   * NOTE: when you do that, probably best to dump everything for the 
   * given language and reconnect as if from scratch.
   */
  private FlushQueueOnError(error_message = "pipe error"){
    Object.keys(this.pending_).forEach(key => {
      if( this.pending_[key] ){
        this.pending_[key].reject(error_message);
        this.pending_[key] = null;
        delete this.pending_[key];
      }
    });
    this.queue_.forEach(item => {
      item.reject(error_message);
    })
    this.queue_ = [];
  }

  private ControlMessageCallback(message){
    this.control_messages_.next(message);    
  }

  private ConsoleCallback(response) {
    let obj = response.getConsole();
    switch (obj.getMessageCase()) {
      case messages.Console.MessageCase.PROMPT:
        this.console_messages_.next({ id:response.getId(), type: ConsoleMessageType.PROMPT, text: obj.getPrompt() });
        break;
      
      case messages.Console.MessageCase.TEXT:
        this.console_messages_.next({ id:response.getId(),type: ConsoleMessageType.TEXT, text: obj.getText() });
        break;
      
      case messages.Console.MessageCase.ERR:
        this.console_messages_.next({ id:response.getId(),type: ConsoleMessageType.ERR, text: obj.getErr() });
        break;

      case messages.Console.MessageCase.MIME_DATA:
        this.console_messages_.next({ id:response.getId(),type: ConsoleMessageType.MIME_DATA, mime_type: obj.getMimeData().getMimeType(), data: obj.getMimeData().getData_asU8() });
        break;

      case messages.Console.MessageCase.HISTORY:
        if( this.history_callback_ ){
          this.history_callback_(obj.toObject()).then( history => {

            let response = new messages.CallResponse();
            let variable = MessageUtilities.ObjectToVariable(history);
            response.setResult(variable);

            let data = response.serializeBinary();
            let frame_length = new Int32Array(1);
            let frame = new Uint8Array(data.length + 4);
        
            frame_length[0] = data.length;
            frame.set(new Uint8Array(frame_length.buffer), 0);
            frame.set(data, 4);
        
            this.client_.write(Buffer.from(frame as any)); // ts type is wrong?
          });
        };            
        break;

      case messages.Console.MessageCase.GRAPHICS:

        // graphics messages are specific to R (at least for now), but they
        // are somewhat special in that they may require a response (usually
        // for measuring text). [...]

        // we'll start simple.
        if( this.graphics_message_handler_ ){
          let response = this.graphics_message_handler_(obj); // obj.toObject().graphics);
          if( response ){

            // console.info( "TX response" );

            let data = response.serializeBinary();
            let frame_length = new Int32Array(1);
            let frame = new Uint8Array(data.length + 4);
        
            frame_length[0] = data.length;
            frame.set(new Uint8Array(frame_length.buffer), 0);
            frame.set(data, 4);
        
            this.client_.write(Buffer.from(frame as any)); // ts type is wrong?

          }
        }

        break;
    }
  }

  private Reject(pending:QueuedCommand, message:any){
    let id = pending.id;
    this.pending_[id] = null; // delete?
    delete this.pending_[id];
    setImmediate(() => pending.reject(message));
  }

  private Resolve(pending:QueuedCommand, message:any){
    let id = pending.id;
    this.pending_[id] = null; // delete?
    delete this.pending_[id];
    setImmediate(() => pending.resolve(message));
  }

  private HandleData(data) {

    let resolve = false;
    try {

      let array = new Uint8Array(data.buffer);
      let stack = [];

      // FIXME: there may be partial packets, in that case we need 
      // to keep a buffer around

      while (array && array.length) {
        let byte_length = new Int32Array(array.buffer.slice(0, 4))[0];
        let response = messages.CallResponse.deserializeBinary(array.slice(4, byte_length + 4));
        stack.push(response);
        array = array.slice(byte_length + 4);
      }

      stack.forEach(response => {

        let id = response.getId();
        let pending = this.pending_[id];

        // there are a couple of different cases in which we get a prompt.
        // on startup, there's a first prompt. this arrives along with 
        // (hopefully after) any startup banners.
        
        // next there's a prompt after a command. we treat prompts as the 
        // "result" of console commands, so they resolve the pending 
        // transaction.

        // there's one additional case, where a remote (Excel) command
        // invokes a browser (such as the debugger). in that case we'll 
        // receive an unexpected prompt which takes over execution. we 
        // (the shell) are still sitting on an open prompt.

        // we now handle that via a special "reset" console message that
        // will show up when we exit the browser. this then becomes the 
        // response to the last unexpected prompt...
        
        // [FIXME: need to do this at each level]

        switch (response.getOperationCase()) {

          case messages.CallResponse.OperationCase.FUNCTION_CALL:
            let system_command = response.getFunctionCall().getFunction();
            if( system_command === "reset-prompt" && pending ){
              this.Resolve(pending, -1);
              resolve = true;
            }
            else setImmediate(() => this.ControlMessageCallback(system_command));
            break;

          case messages.CallResponse.OperationCase.CONTROL_MESSAGE:
            let control_message = response.getControlMessage();
            if( control_message === "reset-prompt" && pending ){
              this.Resolve(pending, -1);
              resolve = true;
            }
            else setImmediate(() => this.ControlMessageCallback(control_message));
            break;

          case messages.CallResponse.OperationCase.CONSOLE:
            let message_case = response.getConsole().getMessageCase();

            if (pending && (message_case === messages.Console.MessageCase.PROMPT)) {
              this.Resolve(pending, response.getConsole().getPrompt());
              resolve = true;
            }
            else setImmediate(() => this.ConsoleCallback(response));
            break;

          case messages.CallResponse.OperationCase.ERR:
            if(pending) this.Reject(pending, response.getErr());
            resolve = true;
            break;

          case messages.CallResponse.OperationCase.RESULT:
            let result = MessageUtilities.VariableToObject(response.getResult());
            if(pending) this.Resolve(pending, result);
            resolve = true;
            break;

          default:
            if(pending) this.Resolve(pending, null);
            resolve = true;

        }
      });
    }
    catch (x) {
      console.error(x);
    }

    if (resolve) {
      setImmediate(() => {
        this.SetBusy(false);
        if (!this.busy) this.ProcessQueue()
      });
    }
  };

  private HandleError(err: string | Error) {
    console.error(err);
    this.FlushQueueOnError("pipe error");
  }

  /** clean shutdown, notify the service */
  Close(): Promise<any> {

    // flush queue
    this.queue_ = [];

    // push
    return this.Control("close");

  }

  RegisterConsoleMessages(){
    return new Promise((resolve, reject) => {
      // register for console messages
      console.info("calling console");
      this.Control("console").then(() => {
        console.info("registered as console client");
        resolve();
      });
    });
  }

  /** initialize and connect to service */
  Init(opts: any = {}) {

    opts = opts || {};
    this.pipe_name_ = opts.pipe_name || "";

    return new Promise((resolve, reject) => {
      console.info("Connecting...");

      let client = net.createConnection({ path: "\\\\.\\pipe\\" + this.pipe_name_ }, () => {
        console.log('connected to service');
        resolve();
      });

      client.on("data", data => this.HandleData(data));
      client.on("error", err => this.HandleError(err));
      client.on("close", () => this.HandleError("closed"));
      client.on('end', () => {
        console.log('disconnected from pipe (xmit)');
      });

      this.client_ = client;

    });

  }

}

File: Common/pipe.cc
/**
 * Copyright (c) 2017-2018 Structured Data, LLC
 * 
 * This file is part of BERT.
 *
 * BERT is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * BERT is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with BERT.  If not, see <http://www.gnu.org/licenses/>.
 */
 
#include "pipe.h"

Pipe::Pipe()
  : buffer_size_(DEFAULT_BUFFER_SIZE)
  , read_buffer_(0)
  , reading_(false)
  , writing_(false)
  , connected_(false)
{
}

Pipe::~Pipe() {
  if (read_buffer_) delete read_buffer_;
}

HANDLE Pipe::pipe_handle() { return handle_; }

HANDLE Pipe::wait_handle_read() { return read_io_.hEvent; }

HANDLE Pipe::wait_handle_write() { return write_io_.hEvent; }

DWORD Pipe::buffer_size() { return buffer_size_; }

void Pipe::QueueWrites(std::vector<std::string> &list) {
  for (auto entry : list) {
    write_stack_.push_back(entry);
  }
}

void Pipe::PushWrite(const std::string &message) {
  write_stack_.push_back(message);
  NextWrite();
}

int Pipe::StartRead() {
  if (reading_ || error_ || !connected_) return 0;
  reading_ = true;
  ResetEvent(read_io_.hEvent);
  ReadFile(handle_, read_buffer_, buffer_size_, 0, &read_io_);
  return 0;
}

void Pipe::ClearError() {
  error_ = false;
}

void Pipe::Connect(bool start_read) {
  connected_ = true;
  std::cout << "pipe connected (" << name_ << ")" << std::endl;
  if (start_read) StartRead();
}

DWORD Pipe::Reset() {

  CancelIo(handle_);
  ResetEvent(read_io_.hEvent);
  ResetEvent(write_io_.hEvent);
  DisconnectNamedPipe(handle_);

  connected_ = reading_ = writing_ = error_ = false;
  message_buffer_.clear();

  if (ConnectNamedPipe(handle_, &read_io_)) {
    std::cerr << "ERR in connectNamedPipe" << std::endl;
  }
  else {
    switch (GetLastError()) {
    case ERROR_PIPE_CONNECTED:
      SetEvent(read_io_.hEvent);
      break;
    case ERROR_IO_PENDING:
      break;
    default:
      std::cerr << "connect failed with " << GetLastError() << std::endl;
      return GetLastError();
      break;
    }
  }

  return 0;
}

int Pipe::NextWrite() {

  DWORD bytes, result;

  if (write_stack_.size() == 0) {
    ResetEvent(write_io_.hEvent);
    writing_ = false;
    return 0; // no write
  }
  else if (writing_) {

    // if we are currently in a write operation, 
    // check if it's complete.

//    result = GetOverlappedResultEx(handle_, &write_io_, &bytes, 0, FALSE);
    result = GetOverlappedResult(handle_, &write_io_, &bytes, FALSE);
    if (!result) {
      DWORD err = GetLastError();
      switch (err) {
      case WAIT_TIMEOUT:
      case WAIT_IO_COMPLETION:
      case ERROR_IO_INCOMPLETE:
        return 0; // write in progress
      default:
        error_ = true;
        std::cout << "Pipe err " << err << std::endl;
        return 0; // err (need a flag here)
      }
    }
  }

  // at this point we're safe to write. we can push more than
  // one write if they complete immediately

  while (write_stack_.size()) {

    ResetEvent(write_io_.hEvent);
    std::string message = write_stack_.front();
    write_stack_.pop_front();
    WriteFile(handle_, message.c_str(), (DWORD)message.length(), NULL, &write_io_);
    //result = GetOverlappedResultEx(handle_, &write_io_, &bytes, 0, FALSE);
    result = GetOverlappedResult(handle_, &write_io_, &bytes, FALSE);
    if (result) {
      // std::cout << "immediate write success (" << bytes << ")" << std::endl;
      ResetEvent(write_io_.hEvent);
      writing_ = false;
    }
    else {
      // std::cout << "write pending" << std::endl;
      writing_ = true;
      return 0;
    }
  }

  return 1;
}

DWORD Pipe::Read(std::string &buf, bool block) {

  // FIXME: what happens in message mode when the buffer is too small? 
  //
  // Docs:
  // In this situation, GetLastError returns ERROR_MORE_DATA, and the client can read the 
  // remainder of the message using additional calls to ReadFile.
  // 

  // that implies we will need to hold on to the buffer.

  DWORD bytes = 0;
//  DWORD success = GetOverlappedResultEx(handle_, &read_io_, &bytes, block ? INFINITE : 0, FALSE);
  DWORD success = GetOverlappedResult(handle_, &read_io_, &bytes, block ? TRUE : FALSE);

  if (success) {
    if (message_buffer_.length()) {
      // std::cout << "appending long mesage (" << bytes << "), this is the end" << std::endl;
      buf = message_buffer_;
      buf.append(read_buffer_, bytes);
      message_buffer_.clear();
    }
    else buf.assign(read_buffer_, bytes);
    reading_ = false;
    return 0;
  }
  else {
    DWORD err = GetLastError();
    if (err == ERROR_MORE_DATA) {
      // std::cout << "appending long mesage (" << bytes << "), this is a chunk" << std::endl;
      message_buffer_.append(read_buffer_, bytes);
      reading_ = false;
      StartRead();
      return err;
    }
    if (err != WAIT_TIMEOUT) error_ = true;
    return err;
  }

}

std::string Pipe::full_name() {
  std::stringstream pipename;
  pipename << "\\\\.\\pipe\\" << name_;
  return pipename.str().c_str();
}

DWORD Pipe::Start(std::string name, bool wait) {

  name_ = name;

  handle_ = CreateNamedPipeA(full_name().c_str(),
    PIPE_ACCESS_DUPLEX | FILE_FLAG_OVERLAPPED,
    PIPE_TYPE_MESSAGE | PIPE_READMODE_MESSAGE | PIPE_WAIT,
    MAX_PIPE_COUNT,
    DEFAULT_BUFFER_SIZE,
    DEFAULT_BUFFER_SIZE,
    100,
    NULL);

  if (NULL == handle_ || handle_ == INVALID_HANDLE_VALUE) {
    std::cerr << "create pipe failed" << std::endl;
    return -1;
  }

  read_io_.hEvent = CreateEvent(0, TRUE, FALSE, 0);
  write_io_.hEvent = CreateEvent(0, TRUE, FALSE, 0);

  read_buffer_ = new char[buffer_size_];
  
  if (ConnectNamedPipe(handle_, &read_io_)) {
    std::cerr << "ERR in connectNamedPipe" << std::endl;
  }
  else {
    switch (GetLastError()) {
    case ERROR_PIPE_CONNECTED:
      SetEvent(read_io_.hEvent);
      break;
    case ERROR_IO_PENDING:
      break;
    default:
      std::cerr << "connect failed with " << GetLastError() << std::endl;
      return GetLastError();
      break;
    }
  }

  if (!wait) return 0;

  while (true) {

    DWORD bytes;
    //DWORD rslt = GetOverlappedResultEx(handle_, &read_io_, &bytes, 1000, FALSE);

    WaitForSingleObject(handle_, 1000);
    DWORD rslt = GetOverlappedResult(handle_, &read_io_, &bytes, FALSE);


    if (rslt) {
      std::cout << "connected (" << name << ")" << std::endl;
      return 0;
    }
    else {
      DWORD err = GetLastError();
      if (err != WAIT_TIMEOUT) {
        error_ = true;
        return err;
      }
    }

  }


  return -2;
}



